### ISSUE TITLE
TestLOBAppend test failing

### ISSUE THREAD

**Issue by SonalMahambrey1** [godror/godror#254]
Recently started facing this issue with v0.34.0

 go test -v -run TestLOBAppend ./...
eDSN:
Using default database for tests:  /github.com/godror/godror/contrib/free.db/env.sh
export TNS_ADMIN="/github.com/godror/godror/contrib/free.db"
export GODROR_TEST_DSN="user=test password=r97oUPimsmTOIcBaeeDF connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-580592118 connectionClass=TestClassName enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=1\nsysasm=0 sysdba=0 sysoper=0 timezone=UTC"
# user=test password=SECRET-*** connectString=db201911301540_high configDir=/tmp/godror_drv_test-580592118
connectionClass=TestClassName enableEvents=1 externalAuth=0 heterogeneousPool=0
libDir= noTimezoneCheck=0 poolIncrement=2 poolMaxSessions=16 poolMinSessions=2
poolSessionMaxLifetime=5m0s poolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0
standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0 timezone=UTC
Version: v0.34.0
Client: 19.3.0.0.0 Timezone: Local
Server: 19.17.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.17.0.1.0] Timezone: UTC
=== RUN   TestLOBAppend
    z_lob_test.go:296: Failed to create temporary lob: sql: connection is already closed
--- FAIL: TestLOBAppend (0.39s)

Kindly suggest

**Comment by sudarshan12s** [comment]
I don't see with my setup (server is 21c in my case , client is still 19.3 ) . Can you set this env and give the logs getting printed on console. It could be connection slow issue and getting terminated. Does it happen with local DB where latency is better. 

```

export DPI_DEBUG_LEVEL=92   
go test -v -run TestLOBAppend
```

**Comment by SonalMahambrey1** [comment]
With the same setup by enabling debug logs

[root@5b8955b6f2a2 godror]# export DPI_DEBUG_LEVEL=92
[root@5b8955b6f2a2 godror]# go test -v -run TestLOBAppend
ODPI [03147] 2022-10-12 03:26:13.326: ODPI-C 4.4.1
ODPI [03147] 2022-10-12 03:26:13.326: debugging messages initialized at level 92
eDSN:
Using default database for tests:  /github.com/godror/godror/contrib/free.db/env.sh
export TNS_ADMIN="/github.com/godror/godror/contrib/free.db"
export GODROR_TEST_DSN="user=test password=r97oUPimsmTOIcBaeeDF connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-543490021 connectionClass=TestClassName enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=1\nsysasm=0 sysdba=0 sysoper=0 timezone=UTC"
ODPI [03147] 2022-10-12 03:26:13.333: fn start dpiContext_createWithParams
ODPI [03147] 2022-10-12 03:26:13.333: Context Parameters:
ODPI [03147] 2022-10-12 03:26:13.333:     Oracle Client Config Dir: /tmp/godror_drv_test-543490021
ODPI [03147] 2022-10-12 03:26:13.333: Environment Variables:
ODPI [03147] 2022-10-12 03:26:13.333:     TNS_ADMIN => "/tmp/godror_drv_test-543490021"
ODPI [03147] 2022-10-12 03:26:13.333:     LD_LIBRARY_PATH => "/opt/oracle/instantclient_19_3:"
ODPI [03147] 2022-10-12 03:26:13.333: check module directory
ODPI [03147] 2022-10-12 03:26:13.333: module name is /tmp/go-build3666022721/b001/godror.test
ODPI [03147] 2022-10-12 03:26:13.333: load in dir /tmp/go-build3666022721/b001
ODPI [03147] 2022-10-12 03:26:13.333: load with name /tmp/go-build3666022721/b001/libclntsh.so
ODPI [03147] 2022-10-12 03:26:13.333: load by OS failure: /tmp/go-build3666022721/b001/libclntsh.so: cannot open shared object file: No such file or directory
ODPI [03147] 2022-10-12 03:26:13.333: load with OS search heuristics
ODPI [03147] 2022-10-12 03:26:13.333: load with name libclntsh.so
ODPI [03147] 2022-10-12 03:26:13.339: load by OS successful
ODPI [03147] 2022-10-12 03:26:13.339: validating loaded library
ODPI [03147] 2022-10-12 03:26:13.342: fn end dpiContext_createWithParams -> 0
ODPI [03147] 2022-10-12 03:26:13.342: fn start dpiContext_getClientVersion(0x4d280390)
ODPI [03147] 2022-10-12 03:26:13.342: fn end dpiContext_getClientVersion(0x4d280390) -> 0
ODPI [03147] 2022-10-12 03:26:13.342: fn start dpiContext_initCommonCreateParams(0x4d280390)
ODPI [03147] 2022-10-12 03:26:13.342: fn end dpiContext_initCommonCreateParams(0x4d280390) -> 0
ODPI [03147] 2022-10-12 03:26:13.342: fn start dpiContext_initConnCreateParams(0x4d280390)
ODPI [03147] 2022-10-12 03:26:13.342: fn end dpiContext_initConnCreateParams(0x4d280390) -> 0
ODPI [03147] 2022-10-12 03:26:13.342: fn start dpiConn_create(0x4d280390)
ODPI [03147] 2022-10-12 03:26:14.755: OCI error ORA-28002: the password will expire within 21 days (dpiConn_create / begin session)
ODPI [03147] 2022-10-12 03:26:14.755: fn end dpiConn_create(0x4d280390) -> 0
ODPI [03147] 2022-10-12 03:26:14.756: fn start dpiConn_setCallTimeout(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.756: fn end dpiConn_setCallTimeout(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.756: fn start dpiConn_setCallTimeout(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.756: fn end dpiConn_setCallTimeout(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.756: fn start dpiConn_ping(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.845: fn end dpiConn_ping(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.845: fn start dpiConn_getIsHealthy(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.845: fn end dpiConn_getIsHealthy(0x4d281220) -> 0
# user=test password=SECRET-*** connectString=db201911301540_high configDir=/tmp/godror_drv_test-543490021
connectionClass=TestClassName enableEvents=1 externalAuth=0 heterogeneousPool=0
libDir= noTimezoneCheck=0 poolIncrement=2 poolMaxSessions=16 poolMinSessions=2
poolSessionMaxLifetime=5m0s poolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0
standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0 timezone=UTC
Version: v0.34.0
ODPI [03147] 2022-10-12 03:26:14.845: fn start dpiConn_getIsHealthy(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.845: fn end dpiConn_getIsHealthy(0x4d281220) -> 0
Client: 19.3.0.0.0 Timezone: Local
ODPI [03147] 2022-10-12 03:26:14.845: fn start dpiConn_getServerVersion(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.935: fn end dpiConn_getServerVersion(0x4d281220) -> 0
Server: 19.17.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.17.0.1.0] Timezone: UTC
ODPI [03147] 2022-10-12 03:26:14.935: fn start dpiConn_getIsHealthy(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.935: fn end dpiConn_getIsHealthy(0x4d281220) -> 0
=== RUN   TestLOBAppend
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_getIsHealthy(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_getIsHealthy(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_setModule(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_setModule(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_prepareStmt(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: SQL BEGIN DBMS_LOB.createtemporary(:1, TRUE, DBMS_LOB.SESSION); END;
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_prepareStmt(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiStmt_getInfo(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiStmt_getInfo(0x4d3df020) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiStmt_getBindCount(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiStmt_getBindCount(0x4d3df020) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_setCallTimeout(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_setCallTimeout(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_setCallTimeout(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_setCallTimeout(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiConn_newVar(0x4d281220)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiConn_newVar(0x4d281220) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiStmt_bindByPos(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:14.936: fn end dpiStmt_bindByPos(0x4d3df020) -> 0
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiStmt_execute(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:15.220: OCI error ORA-03106: fatal two-task communication protocol error (dpiStmt_execute / execute)
ODPI [03147] 2022-10-12 03:26:15.220: fn end dpiStmt_execute(0x4d3df020) -> -1
ODPI [03147] 2022-10-12 03:26:15.220: fn start dpiVar_release(0x4d3df0e0)
ODPI [03147] 2022-10-12 03:26:15.220: fn end dpiVar_release(0x4d3df0e0) -> 0
ODPI [03147] 2022-10-12 03:26:15.220: fn start dpiStmt_release(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:15.220: fn end dpiStmt_release(0x4d3df020) -> 0
ODPI [03147] 2022-10-12 03:26:15.220: fn start dpiConn_release(0x4d281220)
ODPI [03147] 2022-10-12 03:26:15.312: fn end dpiConn_release(0x4d281220) -> 0
    z_lob_test.go:296: Failed to create temporary lob: sql: connection is already closed

**Comment by tgulacsi** [comment]
```
ODPI [03147] 2022-10-12 03:26:14.936: fn start dpiStmt_execute(0x4d3df020)
ODPI [03147] 2022-10-12 03:26:15.220: OCI error ORA-03106: fatal two-task communication protocol error (dpiStmt_execute / execute)
```

It did sth for half a second, then failed with ORA-03106.

**Comment by sudarshan12s** [comment]
@SonalMahambrey1 
can you provide some more logs.
Add the following entries in sqlnet.ora. If sqlnet.ora is not present, you can add it. export TNS_ADMIN  env to location where sqlnet.ora is placed.  

```
TRACE_LEVEL_CLIENT=16                                                           
SSL_SERVER_DN_MATCH=yes
```

It will generate trace (*.trc) and incident (sub folder incdir*) folders inside /tmp/mylogs/ , Can you share those files (*.trc) inside them.

example:
```
export TNS_ADMIN=~/Downloads/

cat ~/Downloads/sqlnet.ora
ADR_BASE=/tmp/mylogs
TRACE_LEVEL_CLIENT=16  
```

**Comment by SonalMahambrey1** [comment]
[incdir_1.zip](https://github.com/godror/godror/files/9770785/incdir_1.zip)
[trace.zip](https://github.com/godror/godror/files/9770787/trace.zip)

Attaching the requested logs

**Comment by sudarshan12s** [comment]
> [incdir_1.zip](https://github.com/godror/godror/files/9770785/incdir_1.zip) [trace.zip](https://github.com/godror/godror/files/9770787/trace.zip)
> 
> Attaching the requested logs

Thanks for sharing. Does this issue happen consistently?. Is it possible to try with 21c database Version on ADB.

**Comment by SonalMahambrey1** [comment]
Actually Last week when  I had run this test on the same VM, this test was passing , I have the logs for the same. However recently I am seeing this issue consistently. No change in version

**Comment by sudarshan12s** [comment]
> Actually Last week when I had run this test on the same VM, this test was passing , I have the logs for the same. However recently I am seeing this issue consistently. No change in version

Can you share the NLS_LANG settings at client, is this setting modified recently?

**Comment by SonalMahambrey1** [comment]
Today I can see slightly different behavior, test goes further but gets skipped due to bad connection instead of failing.
Also this is behavior is seen inconsistently with client 19 . With client 21 , test always passed

Please see the output for both client 19 and client 21 on the same plarform

### **With Client 19.16**
[root@917ae86db7f8 godror]# go test -v -run TestLOBAppend ./...
go: downloading github.com/go-logfmt/logfmt v0.5.1
go: downloading github.com/google/go-cmp v0.5.8
go: downloading github.com/go-logr/logr v1.2.3
go: downloading github.com/godror/knownpb v0.1.0
go: downloading github.com/oklog/ulid/v2 v2.0.2
go: downloading golang.org/x/sync v0.0.0-20220513210516-0976fa681c29
go: downloading google.golang.org/protobuf v1.27.1
eDSN:
Using default database for tests:  /github.com/godror/godror/contrib/free.db/env.sh
export TNS_ADMIN="/github.com/godror/godror/contrib/free.db"
export GODROR_TEST_DSN="user=test password=r97oUPimsmTOIcBaeeDF connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-1346210244 connectionClass=TestClassName enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=1\nsysasm=0 sysdba=0 sysoper=0 timezone=UTC"
# user=test password=SECRET-*** connectString=db201911301540_high configDir=/tmp/godror_drv_test-1346210244
connectionClass=TestClassName enableEvents=1 externalAuth=0 heterogeneousPool=0
libDir= noTimezoneCheck=0 poolIncrement=2 poolMaxSessions=16 poolMinSessions=2
poolSessionMaxLifetime=5m0s poolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0
standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0 timezone=UTC
Version: v0.34.0
**Client: 19.16.0.0.0** Timezone: Local
Server: 19.17.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.17.0.1.0] Timezone: UTC
=== RUN   TestLOBAppend
    z_lob_test.go:298: tmp: godror.Lob{Reader:(*godror.dpiLobReader)(0xc000202300), IsClob:false}
    **z_lob_test.go:307: driver: bad connection
--- SKIP: TestLOBAppend (1.08s)**
PASS
ok      github.com/godror/godror        2.550s
?       github.com/godror/godror/check  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/dsn    0.003s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/num    0.012s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/sid    0.005s [no tests to run]
[root@917ae86db7f8 godror]# uname -a
Linux 917ae86db7f8 4.18.0-305.49.1.el8_4.x86_64 #1 SMP Wed May 11 09:47:48 EDT 2022 **x86_64 x86_64 x86_64 GNU/Linux**

###  **client 21**
[root@7f29b33cbc64 godror]# go test -v -run TestLOBAppend ./...
eDSN:
Using default database for tests:  /github.com/godror/godror/contrib/free.db/env.sh
export TNS_ADMIN="/github.com/godror/godror/contrib/free.db"
export GODROR_TEST_DSN="user=test password=r97oUPimsmTOIcBaeeDF connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-812403146 connectionClass=TestClassName enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=1\nsysasm=0 sysdba=0 sysoper=0 timezone=UTC"
# user=test password=SECRET-*** connectString=db201911301540_high configDir=/tmp/godror_drv_test-812403146
connectionClass=TestClassName enableEvents=1 externalAuth=0 heterogeneousPool=0
libDir= noTimezoneCheck=0 poolIncrement=2 poolMaxSessions=16 poolMinSessions=2
poolSessionMaxLifetime=5m0s poolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0
standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0 timezone=UTC
Version: v0.34.0
**Client: 21.7.0.0.0** Timezone: Local
Server: 19.17.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.17.0.1.0] Timezone: UTC
=== RUN   TestLOBAppend
    z_lob_test.go:298: tmp: godror.Lob{Reader:(*godror.dpiLobReader)(0xc000016840), IsClob:false}
    z_lob_test.go:322: length: 5
--- **PASS: TestLOBAppend (0.44s)**
PASS
ok      github.com/godror/godror        (cached)
?       github.com/godror/godror/check  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/dsn    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/num    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/godror/godror/sid    (cached) [no tests to run]
[root@7f29b33cbc64 godror]# uname -a
Linux 7f29b33cbc64 4.18.0-305.49.1.el8_4.x86_64 #1 SMP Wed May 11 09:47:48 EDT 2022 **x86_64 x86_64 x86_64 GNU/Linux**

**Comment by sudarshan12s** [comment]
I verified it. The fix is required at 19.17.0.1.0 DB server. Once its available, I shall update. Bug number 33427274 is internally tracked for same.
